src/Form/ConstraintForm.php

```diff
@@ -0,0 +1,286 @@
+      }
+
+      // Skip already added unique constraints.
+      if (!empty($definition['unique']) && $default_type !== $definition['id'] && array_key_exists("{$policy_id}.{$definition['id']}", $constraints)) {
+        continue;
+      }
+
+      $options[$definition['id']] = $definition['name'];
+    }
+
+    $form['policy'] = [
+      '#type' => 'hidden',
+      '#value' => $this->getRequest()->get('password_enhancements_policy')->id(),
+    ];
+
+    $settings_wrapper_id = 'ajax-wrapper';
+    $form['type'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Constraint'),
+      '#description' => $this->t('The type of the constraint.'),
+      '#options' => $options,
+      '#default_value' => $default_type,
+      '#empty_value' => '',
+      '#required' => TRUE,
+      '#ajax' => [
+        'callback' => '::updateForm',
+        'event' => 'change',
+        'wrapper' => $settings_wrapper_id,
+      ],
+      '#disabled' => !$this->entity->isNew(),
+    ];
+
+    $form['required'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Required'),
+      '#description' => $this->t('If the constraint is required then it cannot be marked as optional if the minimum required constraints set by the policy are passed the validation.'),
+      '#default_value' => $this->entity->isRequired(),
+    ];
+
+    $form['wrapper'] = [
+      '#type' => 'container',
+      '#prefix' => "<div id='{$settings_wrapper_id}'>",
+      '#suffix' => '</div>',
+      'descriptionSingular' => [
+        '#type' => 'textfield',
+        '#title' => $this->t('Description (singular)'),
+        '#description' => $this->t('The singular description that should be displayed for the password field.'),
+        '#required' => TRUE,
+      ],
+      'descriptionPlural' => [
+        '#type' => 'textfield',
+        '#title' => $this->t('Description (plural)'),
+        '#description' => $this->t('The plural description that should be displayed for the password field.'),
+        '#required' => TRUE,
+      ],
+      'settings' => [
+        '#type' => 'fieldset',
+        '#title' => $this->t('Constraint settings'),
+        '#tree' => TRUE,
+      ],
+    ];
+
+    try {
+      $form['wrapper'] = $this->updateForm($form, $form_state);
+    }
+    catch (Exception $e) {
+      $this->messenger->addError($this->t('Invalid plugin ID: @error', [
+        '@error' => $e->getMessage(),
+      ]));
+      $this->logger->exception($e);
+    }
+
+    return $form;
+  }
+
+  /**
+   * Updates settings from.
+   *
+   * @param array $form
+   *   The form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The from's current state.
+   *
+   * @return array
+   *   Renderable array for the settings wrapper.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   */
+  public function updateForm(array $form, FormStateInterface $form_state): array {
+    // Get type if exist.
+    if (($type = $this->entity->getType()) === '') {
```

Comment:
How and why this could be an empty string? Should not you enforce its non-emptiness elsewhere?

Reply:
When the entity is new this value is empty, that's why I have to grab it from the $form_state if set, because this method is also can called from the ajax callback.

Reply:
#2 possibly can improve this